### ISSUE TITLE
llama3.3-70b oom need support tp on llama3 model

### ISSUE THREAD

**Issue by fan-niu** [sgl-project/SpecForge#84]
please support tp on llama3 model

**Comment by FlamingoPg** [comment]
Sure, we will support it later.

**Comment by FrankLeeeee** [comment]
@yilian49 do you want to take over this issue for hands-on?

**Comment by fan-niu** [comment]
@FrankLeeeee @FlamingoPg @yilian49 Thanks team, looking forward to this feature release

**Comment by yilian49** [comment]
> [@yilian49](https://github.com/yilian49) do you want to take over this issue for hands-on?

Yes I can take it

**Comment by FrankLeeeee** [comment]
Awesome.

**Comment by fan-niu** [comment]
@yilian49 Hi, is there any progress on this task? Thanks

**Comment by yilian49** [comment]
> [@yilian49](https://github.com/yilian49) Hi, is there any progress on this task? Thanks

Hi, yes PR will be ready by today I think

**Comment by yilian49** [comment]
@FlamingoPg @FrankLeeeee would anyone be able to merge this? Thanks! https://github.com/sgl-project/SpecForge/pull/130

**Comment by fan-niu** [comment]
@yilian49 Cool thanks a lot for this task.

**Comment by fan-niu** [comment]
@yilian49 Hi yilian, already used your branch to do tp training but still get oom issue, could you help check this issue? thanks a lot.  The training on 8xh100 and max position = 16384 on llama3-8B-eagle3.json , --max-length 16384 on training script. Thanks a lot. 

training script 
`#export NCCL_DEBUG=INFO
export CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7

torchrun \
    --standalone \
    --nproc_per_node 8\
    $ROOT_DIR/scripts/train_eagle3_online.py \
    --target-model-path $model_path \
    --draft-model-config $ROOT_DIR/configs/llama3-8B-eagle3.json \
    --train-data-path $data_set \
    --output-dir $ROOT_DIR/outputs/llama_3-8B-eagle3 \
    --num-epochs 10 \
    --batch-size 1 \
    --learning-rate 1e-4 \
    --max-length 16384 \
    --chat-template llama3 \
    --cache-dir $ROOT_DIR/cache \
    --warmup-ratio 0.1 \
    --eval-interval 100000 \
    --save-interval 10 \
    --seed 1234 \
    --dist-timeout 3600000 \
    --tp-size 8
`

**Comment by yilian49** [comment]
> @yilian49 Hi yilian, already used your branch to do tp training but still get oom issue, could you help check this issue? thanks a lot.  The training on 8xh100 and max position = 16384 on llama3-8B-eagle3.json , --max-length 16384 on training script. Thanks a lot. 
> 
> training script 
> `#export NCCL_DEBUG=INFO
> export CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
> 
> torchrun \
>     --standalone \
>     --nproc_per_node 8\
>     $ROOT_DIR/scripts/train_eagle3_online.py \
>     --target-model-path $model_path \
>     --draft-model-config $ROOT_DIR/configs/llama3-8B-eagle3.json \
>     --train-data-path $data_set \
>     --output-dir $ROOT_DIR/outputs/llama_3-8B-eagle3 \
>     --num-epochs 10 \
>     --batch-size 1 \
>     --learning-rate 1e-4 \
>     --max-length 16384 \
>     --chat-template llama3 \
>     --cache-dir $ROOT_DIR/cache \
>     --warmup-ratio 0.1 \
>     --eval-interval 100000 \
>     --save-interval 10 \
>     --seed 1234 \
>     --dist-timeout 3600000 \
>     --tp-size 8
> `

Yes I will take a look and update the code

**Comment by fan-niu** [comment]
@yilian49 Hi I see the code has been merged into main, but llama3.1-8b, 8xh100, tp=8, ctx=16k still has OOM issues. Can you help me check it out? Thanks
